File: backend/app/api/projects.py
from fastapi import APIRouter, HTTPException, Depends, Header
from typing import Optional
import logging
from app.models.schemas import ProjectCreate, ProjectUpdate, Project, ProjectList
from app.core.database import (
    create_project,
    get_user_projects,
    get_project_by_id,
    update_project,
    delete_project,
    get_project_analyses,
    link_analysis_to_project,
    supabase_admin
)

router = APIRouter()
logger = logging.getLogger(__name__)


async def get_current_user(authorization: Optional[str] = Header(None)):
    """
    Extract and verify user from JWT token
    """
    if not authorization:
        logger.error("❌ No authorization header")
        raise HTTPException(status_code=401, detail="Authorization header missing")
    
    try:
        if not authorization.startswith("Bearer "):
            logger.error("❌ Invalid authorization format")
            raise HTTPException(status_code=401, detail="Invalid authorization format")
        
        token = authorization.replace("Bearer ", "")
        
        # Verify token with Supabase
        user_response = supabase_admin.auth.get_user(token)
        
        if not user_response or not user_response.user:
            logger.error("❌ No user in response")
            raise HTTPException(status_code=401, detail="Invalid or expired token")
        
        logger.info(f"✅ Authenticated user: {user_response.user.id}")
        return user_response.user
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"❌ Authentication error: {str(e)}")
        raise HTTPException(status_code=401, detail="Invalid token")


@router.post("/projects", response_model=Project)
async def create_new_project(
    project_data: ProjectCreate,
    current_user = Depends(get_current_user)
):
    """
    Create a new project
    """
    try:
        logger.info(f"📝 Creating project for user {current_user.id}: {project_data.name}")
        
        result = await create_project(
            user_id=str(current_user.id),
            project_name=project_data.name,
            project_description=project_data.description
        )
        
        return Project(
            id=result["id"],
            user_id=result["user_id"],
            name=result["name"],
            description=result.get("description"),
            analysis_count=0,  # New projects have no analyses
            created_at=result["created_at"],
            updated_at=result.get("updated_at")
        )
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"❌ Error creating project: {str(e)}")
        raise HTTPException(status_code=500, detail=f"Error creating project: {str(e)}")


@router.get("/projects", response_model=ProjectList)
async def list_user_projects(
    search: Optional[str] = None,
    limit: int = 100,
    current_user = Depends(get_current_user)
):
    """
    List all projects for the current user
    """
    try:
        logger.info(f"📋 Fetching projects for user {current_user.id}")
        
        projects = await get_user_projects(str(current_user.id), limit=limit)
        
        # Filter by search term if provided
        if search:
            search_lower = search.lower()
            projects = [
                p for p in projects 
                if search_lower in p["name"].lower() or 
                   (p.get("description") and search_lower in p["description"].lower())
            ]
        
        # Convert to response model
        project_list = []
        for p in projects:
            project_list.append(Project(
                id=p["id"],
                user_id=p["user_id"],
                name=p["name"],
                description=p.get("description"),
                analysis_count=p.get("analysis_count", 0),
                created_at=p["created_at"],
                updated_at=p.get("updated_at")
            ))
        
        return ProjectList(projects=project_list, total=len(project_list))
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"❌ Error fetching projects: {str(e)}")
        raise HTTPException(status_code=500, detail=f"Error fetching projects: {str(e)}")


@router.get("/projects/{project_id}")
async def get_project_detail(
    project_id: str,
    current_user = Depends(get_current_user)
):
    """
    Get project details with its analyses
    """
    try:
        logger.info(f"🔍 Fetching project {project_id}")
        
        project = await get_project_by_id(project_id, user_id=str(current_user.id))
        
        if not project:
            raise HTTPException(status_code=404, detail="Project not found")
        
        # Get project analyses
        analyses = await get_project_analyses(project_id)
        
        return {
            "id": project["id"],
            "user_id": project["user_id"],
            "name": project["name"],
            "description": project.get("description"),
            "analysis_count": len(analyses),
            "analyses": analyses,
            "created_at": project["created_at"],
            "updated_at": project.get("updated_at")
        }
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"❌ Error fetching project detail: {str(e)}")
        raise HTTPException(status_code=500, detail=f"Error fetching project: {str(e)}")


@router.put("/projects/{project_id}")
async def update_project_info(
    project_id: str,
    project_data: ProjectUpdate,
    current_user = Depends(get_current_user)
):
    """
    Update project information
    """
    try:
        logger.info(f"✏️ Updating project {project_id}")
        
        # Check if project exists and belongs to user
        project = await get_project_by_id(project_id, user_id=str(current_user.id))
        
        if not project:
            raise HTTPException(status_code=404, detail="Project not found")
        
        # Prepare update data
        update_data = {}
        if project_data.name is not None:
            update_data["name"] = project_data.name
        if project_data.description is not None:
            update_data["description"] = project_data.description
        
        if not update_data:
            raise HTTPException(status_code=400, detail="No fields to update")
        
        result = await update_project(project_id, str(current_user.id), **update_data)
        
        return {
            "message": "Project updated successfully",
            "project": result
        }
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"❌ Error updating project: {str(e)}")
        raise HTTPException(status_code=500, detail=f"Error updating project: {str(e)}")


@router.delete("/projects/{project_id}")
async def delete_project_endpoint(
    project_id: str,
    current_user = Depends(get_current_user)
):
    """
    Delete a project and its analyses
    """
    try:
        logger.info(f"🗑️ Deleting project {project_id}")
        
        # Check if project exists and belongs to user
        project = await get_project_by_id(project_id, user_id=str(current_user.id))
        
        if not project:
            raise HTTPException(status_code=404, detail="Project not found")
        
        success = await delete_project(project_id, str(current_user.id))
        
        if not success:
            raise HTTPException(status_code=500, detail="Failed to delete project")
        
        return {"message": "Project deleted successfully"}
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"❌ Error deleting project: {str(e)}")
        raise HTTPException(status_code=500, detail=f"Error deleting project: {str(e)}")


@router.post("/projects/{project_id}/analyses/{analysis_id}")
async def link_analysis_to_project_endpoint(
    project_id: str,
    analysis_id: str,
    current_user = Depends(get_current_user)
):
    """
    Link an analysis to a project
    """
    try:
        logger.info(f"🔗 Linking analysis {analysis_id} to project {project_id}")
        
        # Check if project exists and belongs to user
        project = await get_project_by_id(project_id, user_id=str(current_user.id))
        
        if not project:
            raise HTTPException(status_code=404, detail="Project not found")
        
        success = await link_analysis_to_project(analysis_id, project_id, str(current_user.id))
        
        if not success:
            raise HTTPException(status_code=500, detail="Failed to link analysis")
        
        return {"message": "Analysis linked to project successfully"}
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"❌ Error linking analysis: {str(e)}")
        raise HTTPException(status_code=500, detail=f"Error linking analysis: {str(e)}")

File: frontend/src/services/projects.js
import api from './api';

const PROJECT_ENDPOINTS = {
  CREATE: '/projects',
  LIST: '/projects',
  GET: (projectId) => `/projects/${projectId}`,
  UPDATE: (projectId) => `/projects/${projectId}`,
  DELETE: (projectId) => `/projects/${projectId}`,
  LINK_ANALYSIS: (projectId, analysisId) => `/projects/${projectId}/analyses/${analysisId}`,
};

export const projectService = {
  /**
   * Create a new project
   */
  async createProject(name, description = '') {
    try {
      const response = await api.post(PROJECT_ENDPOINTS.CREATE, {
        name,
        description,
      });
      console.log('✅ Project created:', response.data);
      return response.data;
    } catch (error) {
      console.error('❌ Error creating project:', error);
      throw error;
    }
  },

  /**
   * Get all projects for current user
   */
  async getProjects(search = '', limit = 100) {
    try {
      const response = await api.get(PROJECT_ENDPOINTS.LIST, {
        params: {
          search: search || undefined,
          limit,
        },
      });
      console.log('✅ Projects fetched:', response.data);
      return response.data;
    } catch (error) {
      console.error('❌ Error fetching projects:', error);
      throw error;
    }
  },

  /**
   * Get a specific project with its analyses
   */
  async getProjectDetail(projectId) {
    try {
      const response = await api.get(PROJECT_ENDPOINTS.GET(projectId));
      console.log('✅ Project detail fetched:', response.data);
      return response.data;
    } catch (error) {
      console.error('❌ Error fetching project detail:', error);
      throw error;
    }
  },

  /**
   * Update project information
   */
  async updateProject(projectId, updates) {
    try {
      const response = await api.put(PROJECT_ENDPOINTS.UPDATE(projectId), updates);
      console.log('✅ Project updated:', response.data);
      return response.data;
    } catch (error) {
      console.error('❌ Error updating project:', error);
      throw error;
    }
  },

  /**
   * Delete a project
   */
  async deleteProject(projectId) {
    try {
      const response = await api.delete(PROJECT_ENDPOINTS.DELETE(projectId));
      console.log('✅ Project deleted:', response.data);
      return response.data;
    } catch (error) {
      console.error('❌ Error deleting project:', error);
      throw error;
    }
  },

  /**
   * Link an analysis to a project
   */
  async linkAnalysisToProject(projectId, analysisId) {
    try {
      const response = await api.post(
        PROJECT_ENDPOINTS.LINK_ANALYSIS(projectId, analysisId)
      );
      console.log('✅ Analysis linked to project:', response.data);
      return response.data;
    } catch (error) {
      console.error('❌ Error linking analysis:', error);
      throw error;
    }
  },
};

export default projectService;
